### ISSUE TITLE
mini-css-extract-plugin with SSR.

### ISSUE THREAD

**Issue by princesoni1989** [webpack-contrib/mini-css-extract-plugin#90]
How can we use `mini-css-extract-plugin` with SSR and server bundle.I tried using it is gives error
`document is not defined`

**Comment by evilebottnawi** [comment]
@princesoni1989 looks like a bug/feature. Can you create minimum reproducible example?

**Comment by bwhitty** [comment]
I'd recommend reading https://github.com/webpack-contrib/mini-css-extract-plugin/issues/48#issuecomment-375288454

**Comment by sokra** [comment]
Don't use this plugin in a server bundle. Use `null-loader` or `css-loader/locals` (for CSS modules) instead.

**Comment by johnelliott** [comment]
@sokra just to clarify your suggestion, how would a user import the generated asset name into a bundle to generate a <link ... > tag in a template language or server-side react? 

https://github.com/webpack-contrib/mini-css-extract-plugin#using-preloaded-or-inlined-css

The 'runtime code' aspect of this module really confuses me. I don't understand why there needs to be runtime code for CSS extraction, so I must be trying to use this module in the wrong context. It seems highly geared towards JS-heavy single-page apps that load more JS and CSS asynchronously.

My goal is to server-render a react/redux app with user authentication, and send the user HTML for their logged-in state of the app based on a URL and their credentials. I want to have a <link> tag in HTML that contains the css path on disk with the content hash generated by webpack. Am I on the right track by using this module? The docs for the extract-text-webpack-plugin try to discourage users from using it with webpack 4, but this module doesn't seem to suit some of the past use cases.

**Comment by DimychOcean** [comment]
My experiments with SSR have shown that you can skip the "requireEnsure" hook on the server side. And I came to this solution: override method "getCssChunkObject" in MiniCssExtractPlugin class.

```
class ServerMiniCssExtractPlugin extends MiniCssExtractPlugin {
  getCssChunkObject(mainChunk) {
    return {};
  }
}
```
And then in server webpack config used class ServerMiniCssExtractPlugin.
My rule for CSS in config for server:
```
      {
        test: /\.css$/,
        use: [
          {
            loader: ServerMiniCssExtractPlugin.loader
          },
          {
            loader: 'css-loader',
            options: {
              importLoaders: 1,
              modules: true,
              localIdentName: '[name]__[local]--[hash:base64:5]',
              sourceMap: true
            }
          },
          {
            loader: 'postcss-loader',
            options: {
              plugins: () => [
                require('postcss-flexbugs-fixes'),
                require('autoprefixer')({
                  flexbox: 'no-2009',
                  grid: true
                })
              ]
            }
          }
        ]
      }
```
It's work for me.

**Comment by johnelliott** [comment]
I just skipped this module entirely and I’m now using the css-loader and extract-loader.

**Comment by 132yse** [comment]
I also encountered this problem and thought it was a bug. Would you solve it? Or can I only solve it myself?

**Comment by alejodaraio** [comment]
To me works with the @sokra  solution.

```
appConfig.module.rules.push({
        test: /\.(sa|sc|c)ss$/,
        use: "css-loader/locals"
});
```

**Comment by winterJIE** [comment]
reply
------------------
winter regard.


 




------------------ 原始邮件 ------------------
发件人: "Alejo Daraio"<notifications@github.com>;
发送时间: 2018年8月6日(星期一) 中午11:55
收件人: "webpack-contrib/mini-css-extract-plugin"<mini-css-extract-plugin@noreply.github.com>;
抄送: "Subscribed"<subscribed@noreply.github.com>; 
主题: Re: [webpack-contrib/mini-css-extract-plugin] mini-css-extract-pluginwith SSR. (#90)




To me works with the @sokra  solution.
 appConfig.module.rules.push({         test: /\.(sa|sc|c)ss$/,         use: "css-loader/locals" });

**Comment by michael-ciniawsky** [comment]
https://github.com/webpack-contrib/mini-css-extract-plugin/issues/90#issuecomment-380796867

**Comment by ozguruysal** [comment]
@michael-ciniawsky since you closed the issue and reading the link your pointed, does that mean there'll be no SSR support in this plugin?

**Comment by BODhaha** [comment]
请问你解决了吗？

**Comment by Avcajaraville** [comment]
@michael-ciniawsky Regarding you affirming there will be no SSR for this plugin (you thumbed up @ozguruysal ), don't you guys think this is a bit harsh for people using ssr ? Do you have an alternative we can use ? I read the comments and suggestions and Im not able to get the requirements fulfilled (to extract the css from our vue components in prod. and generate appropriate CSS).

**Comment by evilebottnawi** [comment]
@Avcajaraville Can you create new issue with test repo and describe what you have right now and what you expected and why?

**Comment by Avcajaraville** [comment]
@evilebottnawi Alright, I created a test repo for the issue, with comments on the README.md

Here is the repo: https://github.com/Avcajaraville/debugging-mini-css-extract-plugin

Please, read the readme and let me know if something is not clear.

Thanks a lot for taking time to look into this :)

I cant wait to have this working on my project ;)

**Comment by Avcajaraville** [comment]
@evilebottnawi Did you look at my repo ? Im spending almost every day revisiting the issue and trying to come up with a solution.... no luck 😂

**Comment by 132yse** [comment]
Im spending almost every day revisiting this issue and trying to come up with a solution……no luck，too……

**Comment by Avcajaraville** [comment]
@132yse Yeah, pretty sad to be blocked 😂 

Im considering now to move into Nuxt.js, but I love to control my code and trying to avoid that. The only missing feature I have is to extract my css into single file and per routes. 

Once I have that solved, I can work with a stack of ssr/vue

This evening I will be reviewing Nuxt and how (if) they solve this issue.

Hopefully, one day we get a solution for this :)

**Comment by ozguruysal** [comment]
@Avcajaraville there's a solution to this issue (at least for Vue JS) in my vue-ssr boilerplate. Please review this [webpack config](https://github.com/ozguruysal/vue-ssr-boilerplate/blob/master/build/webpack.base.config.js#L76).  Basically you extract css only from `App.vue` or any other file(s) you specify. You don't need to extract CSS from all components/views because Vue SSR will do the critical CSS injection automatically. See [here](https://ssr.vuejs.org/guide/css.html#css-management). So what you need (splitting css by route) is achieved.

**Comment by 132yse** [comment]
@Avcajaraville Only `createRender` will report an error. If you really want to use Vue ssr, try `CreateBundleRender`, and you won't report an error. The fundamental reason is that `createRender` will have a code :`document.createElement('link)`,this code will create a style link.

**Comment by Avcajaraville** [comment]
@ozguruysal The solution you are posting its not usable for our cases.

Please, find an example of your implementation [here](https://github.com/Avcajaraville/debugging-mini-css-extract-plugin/tree/ozguruysal-solution). You can npm install and run (instruction on readme file).

The main issue is that, as seen on the previously linked example, that I have a route (About) and there I declare styles for body, then I navigate to another (Home), where I also have styles for the body, and then back to About, the styles don't match again.

@132yse I'm not sure I understand how can I can give a note to Vue SSR that says createBundleRender won't report errors because it produces a JSON file that doesn't contain DOM operations. In any case, Im not using createRender but createBundleRender and still hitting the same issue. Please, [check my repo](https://github.com/Avcajaraville/debugging-mini-css-extract-plugin) as well to see what I did.

I tried all of these solutions and still no luck ... :(

**Comment by ozguruysal** [comment]
@Avcajaraville I checked your repo and my suggestion does solve your problem about mini-css-extract plugin. But I see that you have another issue with changing body styles from different components which is not about this plugin. 

There are probably various ways to solve that issue. One is simply changing body class.

**Comment by Avcajaraville** [comment]
@markkelsall None of the proposed solutions work for my case. Still stuck with no success :(

**Comment by arenddeboer** [comment]
I'm in the same boat, unable to get this working with Vue/SSR. I switched back to extract-text-webpack-plugin@2.0.0-beta.4 for now.

**Comment by Avcajaraville** [comment]
@arenddeboer We are a lot of people having same issue, this repo contains already a few related issues and comments. Hope we can get some support eventually. Its really frustrating :|

**Comment by Avcajaraville** [comment]
So, following @sokra suggestion, I changed the config for the server, not using the plugin but null-loader when creating the bundle.

Now my css is properly extracted (yihaaa!), but now all of the css gets in page load, which is not correct.

As an example, on landing page (home) I have also the styles for another route (about).

The JS code splitting works fine, but not the CSS.

Any idea of how to accomplish this ?

As reference, I update the [repo](https://github.com/Avcajaraville/debugging-mini-css-extract-plugin/tree/ozguruysal-solution) with the changes I did.

To sum up; 

this is a screenshot of the assets loaded on the landing page:
![captura de pantalla 2018-11-14 21 34 53](https://user-images.githubusercontent.com/2271201/48510802-2e36af00-e855-11e8-94b2-95d43da8a215.png)

/dist/4.495083d2ddf352f26587.css should not be loaded on page load, but when navigating to /about.

I guess this is probably not a plugin issue but a webpack... Im so lost with this, maybe you guys can help and we can solve the issue together ... :)

**Comment by Laegel** [comment]
Hey guys, thanks for your work and the answers you've already provided. Unfortunately, I'm having the same issue on my project. Any idea why the styles are fetched on the page load and not on demand like the JS? Thanks in advance!

**Comment by Avcajaraville** [comment]
@neverthanmore Nope, this whole webpack and ssr is driving me crazy ! I cant find anything nor Im getting any feedback ...

If you find something, please let me know :) good luck !!

**Comment by evilebottnawi** [comment]
@Avcajaraville please create new issue with minimum reproducible test repo

**Comment by neverthanmore** [comment]
@Avcajaraville when we pack client bundle, we can see css bundle in 
![image](https://user-images.githubusercontent.com/8471969/48626272-6ae8de80-e9ec-11e8-88a8-ab57ab0c821a.png)

so we should move them to async , just like this
![image](https://user-images.githubusercontent.com/8471969/48626360-9ec40400-e9ec-11e8-9253-7a1a9dd12706.png)

these pack logic  is in  node_modules/vue-server-renderer/client-plugin.js
![image](https://user-images.githubusercontent.com/8471969/48626494-f2365200-e9ec-11e8-8e02-dec6ad3414e7.png)

rewrite it
![image](https://user-images.githubusercontent.com/8471969/48626543-1134e400-e9ed-11e8-8b8e-28583b41738b.png)

it works for me! Good luck to u

**Comment by Avcajaraville** [comment]
@neverthanmore Thanks for your suggestion. Unfortunately, this doesn't work for me, and even if it does, it will imply to rewrite a dependency source code, which is not an option for us.

@evilebottnawi Did you manage to have a look to the [issue I opened](https://github.com/webpack-contrib/mini-css-extract-plugin/issues/307) ?

**Comment by chentianyuan** [comment]
@neverthanmore maybe it will be useful [to be released](https://github.com/vuejs/vue/pull/7902)

**Comment by neverthanmore** [comment]
@chentianyuan maybe

**Comment by pulekies** [comment]
For anyone who may run into this issue in the future, updating to the latest version of vue and vue-server-renderer (2.5.21 at this time) along with following @sokra's advice to use _null-loader_ for the server config worked for me.

Updating vue (and vue-server-renderer) also fixed the css issue that @Avcajaraville was running into.

**Comment by chentianyuan** [comment]
@neverthanmore i rm the css in masfest.json from initial to async with a custom webpack plugin， it effective for me

**Comment by neverthanmore** [comment]
@chentianyuan it seems new pr didn't completelty resolve this problem! so sadly

**Comment by 132yse** [comment]
@chentianyuan it seems new pr didn't resolve this problem completely ! so sad

**Comment by GianfrancoCorrea** [comment]
i have same problem, using NextJS in Lerna, its say "window is not defined", no one solved this yet?

**Comment by sebastianks** [comment]
If people still have this issues, consider using this plugin instead: https://github.com/faceyspacey/extract-css-chunks-webpack-plugin ... It just works!

**Comment by cuiyongjian** [comment]
> If people still have this issues, consider using this plugin instead: https://github.com/faceyspacey/extract-css-chunks-webpack-plugin ... It just works!

it works for me. thankyou

**Comment by evilebottnawi** [comment]
Somebody can create minimum reproducible test repo?

**Comment by milewski** [comment]
I fixed the issue by enforcing a limit of 1 single chunk:

```
new webpack.optimize.LimitChunkCountPlugin({
   maxChunks: 1
})
```

**Comment by Avcajaraville** [comment]
Het @evilebottnawi , I did create a repo back in https://github.com/webpack-contrib/mini-css-extract-plugin/issues/90#issuecomment-433336560

Thanks !

**Comment by ScriptedAlchemy** [comment]
mini-css does not impact SSR, you shouldn't be using mini-css as part of an SSR build.
https://github.com/faceyspacey/universal-demo (look at the webpack build) 
What's needed is chunk flushing: https://github.com/faceyspacey/webpack-flush-chunks

The concept is that you utilize the client build to locate assets that need to be served during the first render. 

The mechanism I have designed is implemented in various chunking projects. Though i have not looked outside react at solving the problem. 
https://www.smooth-code.com/open-source/loadable-components/docs/server-side-rendering/

some derivative of webpack-flush-chunks could perform the task

**Comment by nwallis** [comment]
null-loader works for me too.... it prefetches all CSS, but only adds tags for styles used on the page... which is as good as a win for me at this stage.   

In server config:

```
    module: {
        rules: [{
            test: /\.(sa|sc|c)ss$/,
            use: [
                'null-loader',
            ]
        }]
    },
```

**Comment by Murphycx94** [comment]
use extract-css-chunks-webpack-plugin instead mini-css-extract-plugin

webpack.base.config.js
```  javascript
const ExtractCssChunksPlugin = require('extract-css-chunks-webpack-plugin')
{
  ...
  {
        test: /\.css$/,
        use: [
          {
            loader: ExtractCssChunksPlugin.loader,
            options: {
              hot: !isProd,
              reloadAll: !isProd
            }
          },
          'postcss-loader',
          'css-loader'
        ]
      },
      {
        test: /\.less$/,
        use: [
          {
            loader: ExtractCssChunksPlugin.loader,
            options: {
              hot: !isProd,
              reloadAll: !isProd
            }
          },
          'css-loader',
          'postcss-loader',
          'less-loader'
        ]
      }
  ...
  plugins: [
   ...
    new ExtractCssChunksPlugin({
      filename: isProd ? 'css/[name].[contenthash:8].css' : '[name].css',
      chunkFilename: isProd ? 'css/[name].[contenthash:8].chunk.css' : '[name].chunk.css'
    })
  ]
}

```
webpack.server.config.js
``` javascript
{
  ...
  plugins: [
    new webpack.optimize.LimitChunkCountPlugin({
      maxChunks: 1
    })
  ]
}
```
[extract-css-chunks-webpack-plugin](https://www.npmjs.com/package/extract-css-chunks-webpack-plugin)

**Comment by IlyaSemenov** [comment]
So what's the deal with extract-css-chunks-webpack-plugin, how is it different? It tries to 'sell' HMR (even though mini-css-extract-plugin also says HMR is working?), other than that I don't see any difference?

**Comment by ScriptedAlchemy** [comment]
I merged with mini css and assist in maintaining both. Extract is under my total control and I can make more flexible alterations for users. The projects are generally kept in sync

**Comment by slorber** [comment]
> Don't use this plugin in a server bundle. Use null-loader or css-loader/locals (for CSS modules) instead.

If we require regular CSS (not modules) on the server, now that `null-loader` is deprecated, what should be used instead?

> Setting resolve.alias.package to false will tell webpack@5 to ignore a module.

That does not really seem to be applicable to this usecase. 

Creating my own null-loader definitively works, just wondering if it's what I should do 😅

**Comment by alexander-akait** [comment]
We have `emit: true | false` option
